### PR TITLE
C++: Flow through uncertain writes

### PR DESCRIPTION
**Pull request by MathiasVP**
Previously, we weren't getting flow in the following example on the use-use feature branch:
```cpp
char xs[2];
xs[0] = source();
xs[1] = 0;
sink(xs[0]);
```
because the assignment to `xs[1]` was seen as overwriting the contents of `xs`, and thus the assignment `xs[0] = source();` was dead according to SSA.

This PR fixes this by recognizing that some writes are "uncertain" (i.e., we don't know if they completely overwrite a piece of memory), and ensures that we have flow from uncertain definition's prior definition to the variable being defined.

I was actually planning on putting up this PR _after_ https://github.com/github/codeql/pull/11626 had been merged, but seeing as there are some performance issues with that PR I might as well put this on up now (as this PR is largely unrelated to iterators).

### CODE REVIEW

**Comment by aschackmull** [comment]
This seems like a weird use of SSA to me. I'd expect the `char xs[2]` to be the only write to the `xs` pointer, and then the subsequent mentions of `xs` to be reads (the pointer isn't modified). The writes that occur along the way are then store steps in data flow, so nothing is marked as dead by SSA, and there should be no need for uncertain SSA writes.

**Comment by MathiasVP** [comment]
> This seems like a weird use of SSA to me. I'd expect the char xs[2] to be the only write to the xs pointer, and then the subsequent mentions of xs to be reads (the pointer isn't modified). The writes that occur along the way are then store steps in data flow, so nothing is marked as dead by SSA, and there should be no need for uncertain SSA writes.

But we don't model writes to `x[1]` as a store step. We model them as an assignment to an allocation which is tracked by SSA.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll [hunk 1]
```diff
@@ -342,31 +346,60 @@ private predicate indirectConversionFlowStep(Node nFrom, Node nTo) {
...
+    ssaDefinition(input, inp) and
+    defToNode(result, input, _)
+  )
+}
+
+/** Holds if there is def-use or use-use flow from `nodeFrom` to `nodeTo`. */
+predicate ssaFlow(Node nodeFrom, Node nodeTo) {
+  exists(Node nFrom, boolean uncertain |
+    ssaFlowImpl(nFrom, nodeTo, uncertain) and
+    if uncertain = true then nodeFrom = [nFrom, getAPriorDefinition(nFrom)] else nodeFrom = nFrom
```
Comment (rdmarsh2) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:399]: Just to check - this is the only place where we're making a decision based on whether the write is certain, and the rest of the changes are just pushing certainty information through?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/SsaInternals.qll:399]: Yeah, exactly.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**